opentelemetry-api/src/opentelemetry/trace/__init__.py

```diff
@@ -65,8 +65,23 @@
 import typing
 from contextlib import contextmanager
 
+from opentelemetry.trace.status import Status, StatusCanonicalCode
 from opentelemetry.util import loader, types
 
+__all__ = [
```

Comment:
nit: sort alphabetically

Reply:
This is all so we can keep `Status` and `StatusCanonicalCode` in the `opentelemetry.trace` namespace right? If that's the intention then it might not make sense to move them out of this module after all. But why not just use `opentelemetry.trace.status`?

Also AFAICT `__all__` doesn't have any effect here. Since these classes are already in the `opentelemetry.trace` namespace, you'd import them all anyway with the default wildcard import behavior.

Reply:
@reyang, @toumorokoshi, this question is for you.

As I understand it, we only explicitly import submodules in package init files so users don't get `AttributeError`s when they do something like this:

```python
import opentelemetry.sdk
opentelemetry.sdk.trace
```

But in that case we don't need `__all__` to include the submodule names, we just need to import them in the package init file.

It's actually not clear to me why we'd ever use `__all__` except to *prevent* modules or classes from being wildcard-imported. Which isn't how we're using it now in e.g. [`.ext.azure_monitor`](https://github.com/open-telemetry/opentelemetry-python/blob/master/ext/opentelemetry-ext-azure-monitor/src/opentelemetry/ext/azure_monitor/__init__.py#L23) and [`.context.propagation`](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-api/src/opentelemetry/context/propagation/__init__.py).

What does `__all__` give us here?

Reply:
I removed __all__, was having issues with docs creation for this, talked to @lzchen and realized this was not needed, so now opentelemetry.trace.status would be used to access Status and StatusCanonicalCode

Reply:
I agree that it's not needed here since we don't care about hiding the `status` module, but I see now that "preventing imports" exactly what's happening in the azure exporter case. That is, it's intentional that this:

```python
from opentelemetry.ext.azure_monitor import *
```

doesn't import other modules in the package (e.g. `protocol.py`).

So we still have an open question: how much work do we want to put into supporting wildcard imports? I think these `__all__` declarations are going to get out of sync with the project package structure unless we start testing them somehow.

And as an aside: even if we do a bunch of namespace munging in package init files, we're still going to end up exposing the package structure to users. One [weird quirk of the import system](http://python-notes.curiousefficiency.org/en/latest/python_concepts/import_traps.html#the-submodules-are-added-to-the-package-namespace-trap) is that we can't hide internal import paths. E.g. after importing `AzureMonitorSpanExporter` the module `opentelemetry.ext.azure_monitor.protocol` is available. Selectively exposing packages/modules/classes in package init files may be a losing battle.

Reply:
The only good reason I can tell for `__all__` is because people can do `from xyz import *` (for convenience or laziness) and selectively export stuff as intended.
Given we don't allow `from xyz import *`, I don't see a good reason to use `__all__` (if the purpose is to hide some private stuff, we should just use `_` prefix).